runtime/lua/vim/deprecated/health.lua

```diff
@@ -0,0 +1,28 @@
+local M = {}
+local health = vim.health
+
+local deprecated = {}
+
+function M.check()
+  for name, v in vim.spairs(deprecated) do
+    health.start('')
+
+    local alternative, version, backtrace = v[1], v[2], v[3]
+    local major, minor = version:match('(%d+)%.(%d+)')
+    major, minor = tonumber(major), tonumber(minor)
+    local removal_version = string.format('nvim-%d.%d', major, minor)
+    local will_be_removed = vim.fn.has(removal_version) == 1 and 'was removed' or 'will be removed'
+
+    local msg = ('%s is deprecated. Feature %s in Nvim %s'):format(name, will_be_removed, version)
+    local msg_alternative = alternative and ('use %s instead.'):format(alternative)
+    local advice = { msg_alternative }
+    table.insert(advice, backtrace)
+    health.warn(msg, advice)
+  end
+end
+
+function M.add(name, alternative, version, backtrace)
+  deprecated[name] = { alternative, version, backtrace }
```

Comment:
Does it mean that we only collect one backtrace for each deprecated API? I think it would be good to collect multiple invocations. For instance, `vim.tbl_islist` is used in many (20+) places, and one might want to see them all at once rather than `:checkhealth` as many times.